Fix AttributeError when trace is None in genai evaluation

omarfarhoud opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/omarfarhoud/mlflow/pull/19616?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/19616/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/19616/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/19616/merge
```

</p>
</details>

Fixes #19596

## Summary

When using sagemaker-mlflow backend, traces may fail to be retrieved from the tracking store, resulting in `eval_item.trace` being `None`. This causes an `AttributeError` when `_get_new_expectations` tries to access `trace.info.assessments`.

This PR adds defensive null checks in `_get_new_expectations` to handle cases where trace is `None` or `trace.info` is `None`, preventing crashes and allowing evaluation to continue gracefully.

### Related Issues/PRs

Fixes #19596

### What changes are proposed in this pull request?

1. Added null checks at the beginning of `_get_new_expectations` function to handle:
   - Cases where `eval_item.trace` is `None`
   - Cases where `eval_item.trace.info` is `None`
2. Updated function docstring to follow Google-style format and document the defensive behavior
3. Added three unit tests in `tests/genai/evaluate/test_evaluation.py`:
   - `test_get_new_expectations_with_none_trace` - Verifies handling of None trace
   - `test_get_new_expectations_with_none_trace_info` - Verifies handling of None trace.info
   - `test_get_new_expectations_filters_existing_expectations` - Verifies normal filtering behavior still works

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

All tests pass:
```bash
pytest tests/genai/evaluate/test_evaluation.py::TestTraceNoneHandling -v
```

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Fixed `AttributeError` in GenAI evaluation when using backends like SageMaker that don't fully support MLflow tracing. The evaluation harness now gracefully handles cases where traces are unavailable.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [x] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

omarfarhoud committed:
Fix AttributeError when trace is None in genai evaluation

Fixes #19596

When using sagemaker-mlflow backend, traces may fail to be retrieved
from the tracking store, resulting in eval_item.trace being None.
This causes an AttributeError when _get_new_expectations tries to
access trace.info.

Added defensive null checks in _get_new_expectations to handle cases
where trace is None or trace.info is None, preventing crashes and
allowing evaluation to continue gracefully.

Signed-off-by: omarfarhoud <omarfarhoud2003@gmail.com>

WeichenXu123 commented on mlflow/genai/evaluation/harness.py:
```diff
@@ -383,16 +383,35 @@ def run_scorer(scorer):
+        A list of Expectation objects that are new (not already logged to the trace).
+        Returns all expectations from the eval_item if the trace is missing or invalid.
+    """
+    # If trace is missing, skip existing expectation checks
```
```suggestion
    # If trace is missing, try to get expectation assessments from the eval item 
```

WeichenXu123 reviewed (approved):
LGTM.

B-Step62 reviewed (changes requested):
Thank you for filing a PR, @omarfarhoud. However, the `mlflow.genai.evaluation` API fundamentally depends on traces so fixing `_get_new_expectations` is not sufficient.  If the backend does not support tracing, we should raise an exception with a helpful error message that tells users that the backend needs to updated.

ndenStanford commented:
@B-Step62 A gentle bump in case it's buried in your notifications. Appreciate your feedback :).

B-Step62 commented on tests/genai/evaluate/test_evaluation.py:
```diff
@@ -1275,3 +1275,102 @@ def _validate_scorer_max_workers(expected_max_workers, num_scorers):
+    """
+
+    def test_get_new_expectations_with_none_trace(self):
+        """Test that _get_new_expectations handles None trace without crashing.
```
Let's remove single line docstring for test case that does not add much help

B-Step62 commented on tests/genai/evaluate/test_evaluation.py:
```diff
@@ -1275,3 +1275,102 @@ def _validate_scorer_max_workers(expected_max_workers, num_scorers):
+            inputs={"question": "What is the capital of France?"},
+            outputs="Paris",
+            expectations={"expected_response": "Paris"},
+            trace=None,  # Simulates SageMaker backend issue
```
nit: The behavior is not specific to SageMaker. Let's rephrase is to sth like "backend that does not support tracing"

B-Step62 commented on tests/genai/evaluate/test_evaluation.py:
```diff
@@ -1275,3 +1275,102 @@ def _validate_scorer_max_workers(expected_max_workers, num_scorers):
+    Regression test for issue #19596.
+    """
+
+    def test_get_new_expectations_with_none_trace(self):
```
We can reduce the repetitive code by using `pytest.mark.parametrized`

B-Step62 reviewed (commented):
@omarfarhoud Thanks for the reminder, added some comments!

omarfarhoud committed:
resolve conflict

Signed-off-by: B-Step62 <yuki.watanabe@databricks.com>

omarfarhoud commented:
@B-Step62 Thanks for the detailed feedback! I've addressed all your comments:

✅ Converted tests to individual functions (removed class)
✅ Used `pytest.mark.parametrize` to combine similar test cases
✅ Moved imports to top level
✅ Removed verbose single-line docstrings  
✅ Updated comments to be backend-agnostic (not SageMaker-specific)

All tests are passing locally. Ready for another review when you have time!
